{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 193M\r\n",
      "drwxrwxr-x 2 ilya ilya 4.0K Oct 28 16:04 \u001b[0m\u001b[01;34m.\u001b[0m/\r\n",
      "drwxrwxr-x 5 ilya ilya 4.0K Oct 28 16:03 \u001b[01;34m..\u001b[0m/\r\n",
      "-rw-rw-r-- 1 ilya ilya  24M Sep 30 08:54 \u001b[01;31m7E2.pileup.gz\u001b[0m\r\n",
      "-rw-rw-r-- 1 ilya ilya  21M Sep 30 15:49 \u001b[01;31mAAGCTA_R2.pileup.gz\u001b[0m\r\n",
      "-rw-rw-r-- 1 ilya ilya 146M Sep 29 12:04 \u001b[01;31mBJ-HSR1_R1.fastq.gz\u001b[0m\r\n",
      "-rw-rw-r-- 1 ilya ilya 500K Sep 16 09:03 contigs.fasta\r\n",
      "-rw-r----- 1 ilya ilya  644 Sep 16 09:00 dHSR1.fa\r\n",
      "-rw-r----- 1 ilya ilya 1.4M Oct 14 14:41 \u001b[01;31mGCF_000005845.2_ASM584v2_genomic.fna.gz\u001b[0m\r\n",
      "-rw-rw-r-- 1 ilya ilya 440K Sep 10 13:08 \u001b[01;31mGCF_000005845.2_ASM584v2_genomic.gff.gz\u001b[0m\r\n",
      "-rw-rw-r-- 1 ilya ilya 5.4K Sep 23 15:58 gradtimes.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya  445 Sep 16 09:00 hHSR-435.fa\r\n",
      "-rw-rw-r-- 1 ilya ilya  611 Sep 16 09:00 hHSR.fa\r\n",
      "-rw-rw-r-- 1 ilya ilya 4.0K Oct 28 16:04 ROSE1_25.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya 4.2K Oct 28 16:04 ROSE1_26.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya 4.3K Oct 28 16:04 ROSE1_27.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya 4.4K Oct 28 16:04 ROSE1_28.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya 4.5K Oct 28 16:04 ROSE1_29.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya 4.7K Oct 28 16:04 ROSE1_30.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya 5.0K Oct 28 16:04 ROSE1_31.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya 5.2K Oct 28 16:04 ROSE1_32.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya 5.7K Oct 28 16:04 ROSE1_33.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya 6.0K Oct 28 16:04 ROSE1_34.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya 6.5K Oct 28 16:04 ROSE1_35.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya 6.7K Oct 28 16:04 ROSE1_36.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya 7.1K Oct 28 16:04 ROSE1_37.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya 7.8K Oct 28 16:04 ROSE1_38.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya 8.1K Oct 28 16:04 ROSE1_39.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya 8.2K Oct 28 16:04 ROSE1_40.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya 8.6K Oct 28 16:04 ROSE1_41.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya 8.8K Oct 28 16:04 ROSE1_42.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya 9.3K Oct 28 16:04 ROSE1_43.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya  10K Oct 28 16:04 ROSE1_44.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya  11K Oct 28 16:04 ROSE1_45.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya  11K Oct 28 16:04 ROSE1_46.txt\r\n",
      "-rw-rw-r-- 1 ilya ilya  126 Sep 16 09:04 rose.fa\r\n"
     ]
    }
   ],
   "source": [
    "ls -lah ../data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# %load ../../pyRNAfold/utils.py\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from ggplot import *\n",
    "\n",
    "def compute_prob_vector(ps_file, prob_paired=True):\n",
    "    '''\n",
    "    Given a text file derived from the RNAfold output of the form\n",
    "    \n",
    "    i    j    sqrt(prob)   ubox\n",
    "    \n",
    "    computes a vector (dict) of probabilities for every nucleotide\n",
    "    to be in paired (or unpaired) state.\n",
    "    '''\n",
    "    prob_vector = {}\n",
    "    with open(ps_file) as fi:\n",
    "        for line in fi.readlines():\n",
    "            line = line.strip()\n",
    "            posi, posj, sqrt_prob, box = line.split()\n",
    "            curr_i = prob_vector.get(int(posi), 0)\n",
    "            curr_j = prob_vector.get(int(posj), 0)\n",
    "            prob_vector.update({\n",
    "                int(posi): curr_i + float(sqrt_prob)**2,\n",
    "                int(posj): curr_j + float(sqrt_prob)**2,\n",
    "            })\n",
    "    if prob_paired:\n",
    "        return prob_vector\n",
    "    else:\n",
    "        return dict([(pos, 1-p) for pos,p in prob_vector.items()])\n",
    "    \n",
    "\n",
    "def trange_df(base_name, trange=range(35,43), abs_value=True):\n",
    "    '''\n",
    "    Same as `compute_diff_df` but builds dataframe in a long format\n",
    "    suitable for ggplot faceting.\n",
    "    '''\n",
    "    T0 = trange[0]\n",
    "    prob0 = pd.Series(compute_prob_vector('%s_%d.txt' % (base_name, T0)).values())\n",
    "    chunks = []\n",
    "    for temp in trange[1:]:\n",
    "        df = pd.DataFrame()\n",
    "        prob_vector = compute_prob_vector('%s_%d.txt' % (base_name,temp))\n",
    "        df['pos'] = prob_vector.keys()\n",
    "        if abs_value:\n",
    "            df['Diff'] = abs(pd.Series(prob_vector.values()) - prob0)\n",
    "        else:\n",
    "            df['Diff'] = pd.Series(prob_vector.values()) - prob0\n",
    "        df['Temp'] = temp\n",
    "        chunks.append(df)\n",
    "    return pd.concat(chunks)\n",
    "\n",
    "def sig_positions(df, num_sigma=6):\n",
    "    mean = df['Diff'].mean()\n",
    "    sigma = df['Diff'].std()\n",
    "    threshold = num_sigma * sigma\n",
    "    return abs(df['Diff'] - mean) > threshold\n",
    "\n",
    "\n",
    "def compute_diff_df(base_name, trange=range(35,43), abs_value=True):\n",
    "    '''\n",
    "    Given the base_name for tab-delimited files containing base\n",
    "    pairing probabilities calculated by RNAfold computes a\n",
    "    dataframe containing probability difference vectors for each\n",
    "    temperature value in the range relative to the lowest T in the\n",
    "    range.\n",
    "    '''\n",
    "    T0 = trange[0]\n",
    "    prob = compute_prob_vector('%s_%d.txt' % (base_name, T0))\n",
    "    df = pd.DataFrame(prob.items(), columns=['Position', 'Prob_%d' % T0])\n",
    "    for temp in trange[1:]:\n",
    "        prob = compute_prob_vector('%s_%d.txt' % (base_name, temp))\n",
    "        prob_key = 'Prob_%d' % temp\n",
    "        df[prob_key] = pd.Series(prob.values())\n",
    "        if abs_value:\n",
    "            df['Diff_%d' % temp] = abs(df[prob_key] - df['Prob_%d' % T0])\n",
    "        else:\n",
    "            df['Diff_%d' % temp] = df[prob_key] - df['Prob_%d' % T0]\n",
    "    return df\n",
    "\n",
    "def get_sig_positions(df, trange=range(37,43), num_sigma=6):\n",
    "    '''\n",
    "    Given the dataframe of probability differences for a T range\n",
    "    and the level of significannce in sigmas returns positions in the\n",
    "    dataframe where the probability difference at the highest T\n",
    "    exceeds the sigma threshold.\n",
    "    '''\n",
    "    colnames = ['Diff_%d' % temp for temp in trange[1:]]\n",
    "    diff_cols = [df[colname] for colname in colnames]\n",
    "    all_diff = pd.concat(diff_cols)\n",
    "    mean = all_diff.mean()\n",
    "    sigma = all_diff.std()\n",
    "    threshold = num_sigma * sigma\n",
    "    print 'Mean:\\t%f\\nSigma:\\t%f\\nThreshold:\\t%f\\n' % (mean, sigma, threshold)\n",
    "    return df[abs(df['Diff_%d' % trange[-1]] - mean) > threshold].sort(['Position'])\n",
    "\n",
    "def plot_RSMD(df, trange=range(37,43)):\n",
    "   df_sum = pd.DataFrame()\n",
    "   df_sum['Temp'] = trange[1:]\n",
    "   df_sum['RMSD'] = [np.sqrt(((df[df['Temp'] == T]['Diff'])**2).sum()) for T in trange[1:]]\n",
    "   p = ggplot(df_sum, aes(x='Temp', y='RMSD')) + geom_line()\n",
    "   return p\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
